homeassistant/components/tuya/fan.py

```diff
@@ -0,0 +1,267 @@
+
+# Air Purifier
+# https://developer.tuya.com/en/docs/iot/s?id=K9gf48r41mn81
+DPCODE_AP_FAN_SPEED = "speed"
+DPCODE_AP_FAN_SPEED_ENUM = "fan_speed_enum"
+
+TUYA_SUPPORT_TYPE = {
+    "fs",  # Fan
+    "kj",  # Air Purifier
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+):
+    """Set up tuya fan dynamically through tuya discovery."""
+    _LOGGER.debug("fan init")
+
+    hass.data[DOMAIN][entry.entry_id][TUYA_HA_TUYA_MAP][
+        DEVICE_DOMAIN
+    ] = TUYA_SUPPORT_TYPE
+
+    async def async_discover_device(dev_ids):
+        """Discover and add a discovered tuya fan."""
+        _LOGGER.debug("fan add-> %s", dev_ids)
+        if not dev_ids:
+            return
+        entities = await hass.async_add_executor_job(
+            _setup_entities, hass, entry, dev_ids
+        )
+        async_add_entities(entities)
+
+    async_dispatcher_connect(
+        hass, TUYA_DISCOVERY_NEW.format(DEVICE_DOMAIN), async_discover_device
+    )
+
+    device_manager = hass.data[DOMAIN][entry.entry_id][TUYA_DEVICE_MANAGER]
+    device_ids = []
+    for (device_id, device) in device_manager.device_map.items():
+        if device.category in TUYA_SUPPORT_TYPE:
+            device_ids.append(device_id)
+    await async_discover_device(device_ids)
+
+
+def _setup_entities(hass: HomeAssistant, entry: ConfigEntry, device_ids: list):
+    """Set up Tuya Fan."""
+    device_manager = hass.data[DOMAIN][entry.entry_id][TUYA_DEVICE_MANAGER]
+    entities = []
+    for device_id in device_ids:
+        device = device_manager.device_map[device_id]
+        if device is None:
+            continue
+        entities.append(TuyaHaFan(device, device_manager))
+        hass.data[DOMAIN][entry.entry_id][TUYA_HA_DEVICES].add(device_id)
+    return entities
+
+
+class TuyaHaFan(TuyaHaEntity, FanEntity):
+    """Tuya Fan Device."""
+
+    def __init__(self, device: TuyaDevice, device_manager: TuyaDeviceManager) -> None:
+        """Init Tuya Fan Device."""
+        super().__init__(device, device_manager)
+
+        # Air purifier fan can be controlled either via the ranged values or via the enum.
+        # We will always prefer the enumeration if available
+        #   Enum is used for e.g. MEES SmartHIMOX-H06
+        #   Range is used for e.g. Concept CA3000
+        self.air_purifier_speed_range_len = 0
+        self.air_purifier_speed_range_enum = []
+        if self.tuya_device.category == "kj":
+            try:
+                if (
+                    DPCODE_AP_FAN_SPEED_ENUM in self.tuya_device.status
+                    or DPCODE_AP_FAN_SPEED in self.tuya_device.status
+                ):
+
+                    self.dp_code_speed_enum = (
+                        DPCODE_AP_FAN_SPEED_ENUM
+                        if DPCODE_AP_FAN_SPEED_ENUM in self.tuya_device.status
+                        else DPCODE_AP_FAN_SPEED
+                    )
+                    data = json.loads(
```

Comment:
Why do we need to load from json string? This seems like a low level protocol that should be handled by the library.

Reply:
Not all device function values is json string, some of them are boolean, int, so it will be handle in the platform.

Reply:
My point is that the library should return a usable format from the beginning. The library user should not need to convert from a JSON string, since that's most likely not how the library user wants to consume the data.